### PR TITLE
Allow system-wide installs to specify both a root_path and a user.

### LINKED ISSUE #247: It should be possible to use the rbenv_script resource as a non-root user, with a system-wide installation
### Cookbook version
2.1.2

### Chef-client version
14.x

### Platform Details
ubuntu-18.04 (bionic)
ubuntu-14.04 (trusty)

### Scenario:
I use the system-wide recipe to install rbenv and rubies.

I'd like to run an "rbenv-aware" command (even bundler!) as a non-root user and have those commands run as the non-root user, using the system-wide rbenv/ruby install.

This does not seem possible.

### Steps to Reproduce:

First, run the system-wide ruby recipe, and install a ruby. Then, in a recipe:

```ruby
rbenv_script "bundle" do
  user 'deploy'
  group 'deploy'
  cwd '/home/deploy/rails_app'
  code 'bundle'
end
```

It doesn't work, because by specifying a "user", I have also signaled to the cookbook that I have a user-rbenv installation, which I don't (see https://github.com/sous-chefs/ruby_rbenv/blob/9cf17d0e2cbfe8c7844dad265e8fb634b32d7f46/libraries/helpers.rb#L36).

I could do this, to signal that I have a system-wide installation (don't supply a user):

```ruby
rbenv_script "bundle" do
  group 'deploy'
  cwd '/home/deploy/rails_app'
  code 'bundle'
end
```

Bundle runs, but the vendor/bundle directory is owned by root. This is problematic.

### Expected Result:
I expect to be able to run an rbenv_script resource with a system-wide installation as a non-root user.

### Actual Result:
See above.

The "user" resource variable in the "rbenv_script" resource should not be "overloaded" to both signal a user-installation and also be used as the user for the underlying bash resource setup (https://github.com/sous-chefs/ruby_rbenv/blob/9cf17d0e2cbfe8c7844dad265e8fb634b32d7f46/resources/script.rb#L41). By overloading it in this way, we lose the ability to run an rbenv-aware script as a given user but with a system-wide install.

### PR DESCRIPTION
**Pull request by corporealfunk**
### Description

Previously if a user was specified, the cookbook assumed a user rbenv installation and also ran scripts as that user. The only way to signal a system-wide install was to omit a user property for rbenv-* resources. This had unfortunate side effects, for example all script resources would be forced to run as root instead of a specific user.

Allowing one to pass a root_path property to resources that leverage rbenv-script or rbenv-plugin and others required adding a root_path property to almost all the resources.  Because we want to keep the lazy-loaded path DRY, it was necessary to refactor the root_path code into it's own module function (and not an module instance function since we need to call it during resource declaration).

This is not thoroughly tested, though the ubuntu-1804 tests as written pass, using kitchen verify.

I have not tested that specifying a root_path property on all resources results in the root_path property being utilized correctly.

### Issues Resolved

#247 

### Contribution Check List

- [ ] All tests pass.  (kitchen test *-ubuntu-1804 tests pass.)
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/ruby_rbenv/248)

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This Pull Request is probably missing tests.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: resources/gem.rb [hunk 1]
```diff
@@ -35,6 +35,7 @@
 property :response_file,          String # Only used to reconfigure
 property :user,                   String
 property :rbenv_version,          String, required: true
+property :root_path,              String, default: lazy { Chef::Rbenv.root_path(node, user) }
```
Comment (damacus) [inline resources/gem.rb:38]: I was wondering why you were using self above, but this is a really smart pattern. I think we should adopt this across the board.

**Comment by damacus** [comment]
If you can add a CHANGELOG for this I'll squash and merge, and release it.

As noted, the use of self and the namespace is really smart. I think to avoid name collisions we should be using that type of notation in more places.

**Review by damacus — CHANGES_REQUESTED** [review]
Needs a, changelog as per the bots request.

**Pull request closed without merging**